### ISSUE TITLE
Convention for cancelled spans

### ISSUE THREAD

**Issue by charleskorn** [open-telemetry/semantic-conventions#560]
It would be great if there was a language- and protocol-agnostic convention for marking a span as representing a cancelled operation.

At present, different languages and protocols treat cancelled operations in different ways, which makes it difficult to do things like filter a trace to show only non-cancelled operations, or visually distinguish between successful, failed and cancelled operations.

For example, a span representing a gRPC call would have the `rpc.grpc.status_code` attribute set to 1 (for the gRPC `CANCELLED` [status code](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md)), while a span representing a Golang operation might be recorded with an event with a `context canceled` message.

Instead, it'd be great if there was a convention for an attribute that marked a span as cancelled, such as `cancelled`.

**Comment by trask** [comment]
oh no, we'll need to make a decision on canceled vs cancelled 😅

in seriousness, a "canceled" (or "cancelled") attribute would be very welcome in opentelemetry-java-instrumentation repo where we currently have (at least):

* `guava.canceled`
* `rxjava.canceled`
* `jaxrs.canceled`
* `reactor.canceled`
* `grpc.canceled`

**Comment by mateuszrzeszutek** [comment]
Didn't we want to use the `error.type` attribute for that? At least in the HTTP scenarios, e.g. `error.type = cancelled`

**Comment by pyohannes** [comment]
> Didn't we want to use the error.type attribute for that? At least in the HTTP scenarios, e.g. error.type = cancelled

The problem with using `error.type` is, that a cancelled operation doesn't necessarily imply an error. However, the definition of [`error.type`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/error.md) states this:

> If the operation has completed successfully, instrumentations SHOULD NOT set `error.type`.

**Comment by lmolkova** [comment]
I'd be in favor of defining something in common. 

I think we can polish `error.type` description to accommodate it - "If the operation has completed successfully, instrumentations SHOULD NOT set error.type." - cancelled operation did not complete successfully and it should be totally legit to use `error.type` for it. In the same was as we set `error.type` to `"404"` on the client side even when 404 does not represent an error.

Some problem with the `cancel(l)ed` convention:
- E.g. on Java the async operations (like reactor) can be cancelled, but sync versions of them throw something like `TimeoutException` or `InterruptedException` - I'm not sure how far we can go unifying and if by replacing exception type with `cancelled` is helpful.
- There were some concerns about attempts to unify across languages. E.g. in .NET cancellation is commonly done with `OperationCancelledException` that's familar to all .NET devs - what should .net instrumentations do?

The least controversial option I can suggest is to use `cancel(l)ed` when operation is cancelled without exception.

**Comment by pyohannes** [comment]
> cancelled operation did not complete successfully and it should be totally legit to use `error.type`

That depends very much on the context. In some cases, for example if I decide that I don't need the output of an operation anymore, I cancel an operation and cancellation is the desired outcome. I don't want this call to count against failed calls in my downstream calls.

Also, let's make clear whether we are talking about setting `error.type` to `canceled` on spans, on metrics, or on both. For spans, would we additionally set the span status to `ERROR` for canceled operations?

**Comment by lmolkova** [comment]
related to #1712: cancellation is orthogonal to error/success and we need a way to record it on metrics, spans, and events without implying something has failed.

**Comment by trask** [comment]
Merging #1712 into this issue:

> We have an internal request hedging system, similar in concept to what the gRPC team has talked about here: https://grpc.io/docs/guides/request-hedging/
>
> In Golang, we can use `context.WithCancelCause(...)` to cancel a hedge that didn't win.  The winning hedge just looks like a normal Span that was successful.
>
> When we are integrating some 3rd-party OTEL libraries, like [otelpgx](https://github.com/exaring/otelpgx), it uses `span.RecordError(err)` liberally, and so we get Spans marked as "error", when they were really just cancelled.  We can deal with some of this by changing our own software of course.
>
> But when looking at semconv there didn't seem to even be a way to mark an RPC/Database operation as a "hedge" and to denote it was cancelled on purpose.  We've been using some internal terminology like "winning hedge" and "lost hedge".
>
> Has this been discussed before?  I could only find minor references as it being out of scope previously:
https://github.com/open-telemetry/opentelemetry-specification/blob/d1c6aec6524f8c071d5fdef23d57ca5964fa6bb5/oteps/trace/0174-http-semantic-conventions.md?plain=1#L165-L167

**Comment by roji** [comment]
@lmolkova and others, depending on the context and the user's goals, reporting cancellations as errors might not be the right thing. I'm specifically thinking about cases where traced errors are something to be investigated, and possibly even set of an alarm or a red indicator in some dashboard: from that point of view, an operation that was cancelled by the user is not an error at all, but rather a successful "completion" (just in another way, completion without waiting for the operation end etc.).

There may be a distinction to be made here between tracing and metrics (though I'm not sure) - since tracing isn't aggregated and represents actual, single requests, flagging cancellation as error may be more harmful there than doing so with metrics, since there's more of a chance the goal there is to flag actual errors and react to them (a bit like with logging). Whereas for a histogram metrics of execution duration, it seems much more likely that users are interested in, well, the duration, and therefore cancellations are more likely to skew.

In any case, rather than forcing a binary distinction between "success" and "error", where cancellation won't satisfy everyone involved, it may be better to introduce a dedicated status code alongside Error, Ok, Unset; this would mirror the distinction made in various languages, e.g. [.NET's TaskStatus](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskstatus?view=net-9.0).

**Comment by thompson-tomo** [comment]
I Agree with @roji about needing to have a potential different status & as such had created https://github.com/open-telemetry/opentelemetry-specification/issues/4677 which is looking for community feedback.

**Comment by lmolkova** [comment]
RPC SIG discussion:
- additional status idea looks interesting, but we need to be able to make progress on RPC conventions. Eventually we can leverage additional status in instrumentations
- on [HTTP](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#status), we explicitly say that cancellation is an error by default, but if instrumentation has any additional context, it should use that context to set status more accurately

So let's add a similar note to RPC docs and if RPC instrumentations can distinguish user-cancelling / hedging from timeouts, they already can leverage this knowledge

**Comment by lmolkova** [comment]
RPC SIG discussion:
We link to https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/recording-errors.md doc already and have a room to provide additional guidance (if we can have reliable way to classify timeout as an error). For now, it seems there is nothing to do in the context of initial RPC stabilization

**Comment by rkargMsft** [comment]
The challenge I see with this:

> If the HTTP client instrumentation can detect that the request was cancelled intentionally by the caller

is how would an application that makes an HTTP call provide information to the HTTP client instrumentation that a cancellation was intentional?
Specifically looking at a .NET app that calls through HttpClient (hedging) and cancels a bunch of them. The app doesn't handle the HTTP instrumentation (this is generally from a shared package), so it feels like having a standard way to push down that information would be necessary, but I'm not sure how that would actually happen.
